File: ruby/difference-of-squares/difference_of_squares.rb
# frozen_string_literal: true

# Implementation of the Difference of Squares exercise in Ruby track on Exercism.
class Squares
  # @param number [Integer]
  def initialize(number)
    @numbers = (1..number)
  end

  # @return [Integer] the squared sum of all numbers up to and including the given number
  def square_of_sum
    @numbers.sum**2
  end

  # @return [Integer] the sum of squares of all numbers up to and including the given number
  def sum_of_squares
    @numbers.reduce { |sum, i| sum + i**2 }
  end

  # @return [Integer] the difference between the squared sum and sum of squares
  def difference
    square_of_sum - sum_of_squares
  end
end

File: lua/difference-of-squares/difference-of-squares.lua
local function square_of_sum(n)
  sum = 0
  current = n
  while(current > 0)
  do
    sum = sum + current
    current = current - 1
  end
  return sum * sum
end

local function sum_of_squares(n)
  sum = 0
  current = n
  while(current > 0)
  do
    sum = sum + current * current
    current = current - 1
  end
  return sum
end

local function difference_of_squares(n)
  return square_of_sum(n) - sum_of_squares(n)
end

return {
  square_of_sum = square_of_sum,
  sum_of_squares = sum_of_squares,
  difference_of_squares = difference_of_squares
}
